File: src/platform_app/modules/identity/service.py
"""Passwordless activation, sessions, usernames, and principals."""

import hashlib
import hmac
import re
import secrets
from collections.abc import Callable
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from uuid import UUID

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from platform_app.modules.identity.authorization import Principal, RoleGrant
from platform_app.modules.identity.models import (
    AccessSession,
    AccountStatus,
    EmailChallenge,
    EnrollmentState,
    ModerationState,
    Role,
    RoleAssignment,
    University,
    UniversityEmailDomain,
    UniversityIdentity,
    UniversityStatus,
    UserAccount,
    UserProfile,
    VerifiedResidence,
)
from platform_app.modules.identity.ports import EmailCodeSender
from platform_app.modules.university.contracts import (
    UniversityVerificationGateway,
    VerificationStatus,
)
from platform_app.modules.university.contracts import (
    VerifiedResidence as VerifiedResidenceContract,
)

USERNAME_PATTERN = re.compile(r"^[A-Za-z0-9_]{3,24}$")


class IdentityError(ValueError):
    """Safe authentication error base."""


class UniversityDomainError(IdentityError):
    pass


class ChallengeRateLimitError(IdentityError):
    pass


class InvalidChallengeError(IdentityError):
    pass


class RosterIneligibleError(IdentityError):
    pass


class UsernameUnavailableError(IdentityError):
    pass


class InvalidSessionError(IdentityError):
    pass


@dataclass(frozen=True, slots=True)
class ChallengeIssued:
    challenge_id: UUID
    expires_at: datetime


@dataclass(frozen=True, slots=True)
class ActivatedSession:
    access_token: str
    expires_at: datetime
    username: str
    roles: tuple[Role, ...]


class IdentityService:
    """Coordinate passwordless verification and tenant-safe activation."""

    def __init__(
        self,
        session: Session,
        university_gateway: UniversityVerificationGateway,
        email_sender: EmailCodeSender,
        challenge_hmac_key: str,
        session_hmac_key: str,
        *,
        clock: Callable[[], datetime] | None = None,
        code_factory: Callable[[], str] | None = None,
        token_factory: Callable[[], str] | None = None,
    ) -> None:
        self._session = session
        self._university_gateway = university_gateway
        self._email_sender = email_sender
        self._challenge_hmac_key = challenge_hmac_key.encode()
        self._session_hmac_key = session_hmac_key.encode()
        self._clock = clock or (lambda: datetime.now(UTC))
        self._code_factory = code_factory or (
            lambda: f"{secrets.randbelow(1_000_000):06d}"
        )
        self._token_factory = token_factory or (lambda: secrets.token_urlsafe(32))

    def request_challenge(self, email: str) -> ChallengeIssued:
        normalized_email, domain = normalize_email(email)
        university = self._session.scalar(
            select(University)
            .join(
                UniversityEmailDomain,
                UniversityEmailDomain.university_id == University.id,
            )
            .where(
                UniversityEmailDomain.normalized_domain == domain,
                University.status == UniversityStatus.ACTIVE,
            )
        )
        if university is None or university.roster_reference is None:
            raise UniversityDomainError("email is not eligible for participation")

        now = self._now()
        recent_count = self._session.scalar(
            select(func.count(EmailChallenge.id)).where(
                EmailChallenge.normalized_email == normalized_email,
                EmailChallenge.created_at >= now - timedelta(minutes=15),
            )
        )
        if recent_count is not None and recent_count >= 3:
            raise ChallengeRateLimitError("too many verification requests")

        challenge = EmailChallenge(
            university_id=university.id,
            normalized_email=normalized_email,
            code_digest="pending",
            expires_at=now + timedelta(minutes=10),
            attempts=0,
            max_attempts=5,
            created_at=now,
        )
        self._session.add(challenge)
        self._session.flush()
        code = self._code_factory()
        challenge.code_digest = self._digest_challenge(challenge.id, code)
        self._email_sender.send_code(normalized_email, code, challenge.expires_at)
        return ChallengeIssued(challenge.id, challenge.expires_at)

    def verify_challenge(
        self, challenge_id: UUID, code: str, username: str
    ) -> ActivatedSession:
        challenge = self._session.scalar(
            select(EmailChallenge)
            .where(EmailChallenge.id == challenge_id)
            .with_for_update()
        )
        now = self._now()
        if (
            challenge is None
            or challenge.consumed_at is not None
            or challenge.expires_at <= now
            or challenge.attempts >= challenge.max_attempts
        ):
            raise InvalidChallengeError("verification code is invalid")
        expected_digest = self._digest_challenge(challenge.id, code)
        if not hmac.compare_digest(challenge.code_digest, expected_digest):
            challenge.attempts += 1
            self._session.flush()
            raise InvalidChallengeError("verification code is invalid")

        verification = self._university_gateway.verify_resident(
            challenge.normalized_email, now
        )
        university = self._session.get(University, challenge.university_id)
        if (
            university is None
            or verification.status is not VerificationStatus.ACTIVE
            or verification.residence is None
            or verification.student_reference is None
            or verification.university_reference != university.roster_reference
        ):
            challenge.consumed_at = now
            self._session.flush()
            raise RosterIneligibleError("active enrolment and residence are required")

        normalized_username = normalize_username(username)
        identity = self._session.scalar(
            select(UniversityIdentity).where(
                UniversityIdentity.university_id == university.id,
                UniversityIdentity.normalized_email == challenge.normalized_email,
            )
        )
        if identity is None:
            account = UserAccount(status=AccountStatus.ACTIVE)
            self._session.add(account)
            self._session.flush()
            identity = UniversityIdentity(
                university_id=university.id,
                account_id=account.id,
                normalized_email=challenge.normalized_email,
                external_student_reference=verification.student_reference,
                enrollment_state=EnrollmentState.ACTIVE,
            )
            self._session.add(identity)
            self._session.flush()
        else:
            existing_account = self._session.get(UserAccount, identity.account_id)
            if (
                existing_account is None
                or existing_account.status is not AccountStatus.ACTIVE
            ):
                raise RosterIneligibleError("account is not active")
            account = existing_account
            identity.external_student_reference = verification.student_reference
            identity.enrollment_state = EnrollmentState.ACTIVE

        profile = self._upsert_profile(identity, username, normalized_username)
        self._upsert_participant_role(account.id, university.id)
        self._record_residence(identity, verification.residence, now)
        challenge.consumed_at = now
        token = self._token_factory()
        access_session = AccessSession(
            account_id=account.id,
            token_digest=self._digest_session(token),
            expires_at=now + timedelta(hours=1),
            created_at=now,
        )
        self._session.add(access_session)
        self._session.flush()
        return ActivatedSession(
            access_token=token,
            expires_at=access_session.expires_at,
            username=profile.username,
            roles=(Role.PARTICIPANT,),
        )

    def principal_for_token(self, token: str) -> Principal:
        now = self._now()
        access_session = self._session.scalar(
            select(AccessSession).where(
                AccessSession.token_digest == self._digest_session(token),
                AccessSession.revoked_at.is_(None),
                AccessSession.expires_at > now,
            )
        )
        if access_session is None:
            raise InvalidSessionError("access token is invalid")
        assignments = self._session.scalars(
            select(RoleAssignment).where(
                RoleAssignment.account_id == access_session.account_id
            )
        )
        return Principal(
            account_id=access_session.account_id,
            grants=tuple(
                RoleGrant(row.role, row.university_id, row.building_id)
                for row in assignments
            ),
        )

    def change_username(self, principal: Principal, username: str) -> str:
        normalized = normalize_username(username)
        identity = self._session.scalar(
            select(UniversityIdentity).where(
                UniversityIdentity.account_id == principal.account_id
            )
        )
        if identity is None:
            raise InvalidSessionError("participant identity is missing")
        profile = self._session.scalar(
            select(UserProfile).where(UserProfile.identity_id == identity.id)
        )
        if profile is None:
            raise InvalidSessionError("participant profile is missing")
        existing = self._session.scalar(
            select(UserProfile).where(
                UserProfile.university_id == identity.university_id,
                UserProfile.normalized_username == normalized,
                UserProfile.id != profile.id,
            )
        )
        if existing is not None:
            raise UsernameUnavailableError("username is unavailable")
        profile.username = username
        profile.normalized_username = normalized
        self._session.flush()
        return profile.username

    def _upsert_profile(
        self, identity: UniversityIdentity, username: str, normalized_username: str
    ) -> UserProfile:
        profile = self._session.scalar(
            select(UserProfile).where(UserProfile.identity_id == identity.id)
        )
        conflict = self._session.scalar(
            select(UserProfile).where(
                UserProfile.university_id == identity.university_id,
                UserProfile.normalized_username == normalized_username,
                UserProfile.identity_id != identity.id,
            )
        )
        if conflict is not None:
            raise UsernameUnavailableError("username is unavailable")
        if profile is None:
            profile = UserProfile(
                university_id=identity.university_id,
                identity_id=identity.id,
                username=username,
                normalized_username=normalized_username,
                moderation_state=ModerationState.ACTIVE,
            )
            self._session.add(profile)
        else:
            profile.username = username
            profile.normalized_username = normalized_username
        self._session.flush()
        return profile

    def _upsert_participant_role(self, account_id: UUID, university_id: UUID) -> None:
        assignment = self._session.scalar(
            select(RoleAssignment).where(
                RoleAssignment.account_id == account_id,
                RoleAssignment.role == Role.PARTICIPANT,
                RoleAssignment.university_id == university_id,
            )
        )
        if assignment is None:
            self._session.add(
                RoleAssignment(
                    account_id=account_id,
                    role=Role.PARTICIPANT,
                    university_id=university_id,
                    building_id=None,
                )
            )
            self._session.flush()

    def _record_residence(
        self,
        identity: UniversityIdentity,
        residence: VerifiedResidenceContract,
        now: datetime,
    ) -> None:
        current = self._session.scalar(
            select(VerifiedResidence).where(
                VerifiedResidence.identity_id == identity.id,
                VerifiedResidence.effective_end.is_(None),
            )
        )
        if current is not None:
            unchanged = (
                current.building_reference == residence.building_reference
                and current.apartment_reference == residence.apartment_reference
                and current.room_reference == residence.room_reference
                and current.source_version == residence.source_version
            )
            if unchanged:
                current.verified_at = now
                return
            current.effective_end = now
        self._session.add(
            VerifiedResidence(
                university_id=identity.university_id,
                identity_id=identity.id,
                building_reference=residence.building_reference,
                apartment_reference=residence.apartment_reference,
                room_reference=residence.room_reference,
                source_version=residence.source_version,
                effective_start=now,
                effective_end=None,
                verified_at=now,
            )
        )

    def _digest_challenge(self, challenge_id: UUID, code: str) -> str:
        return hmac.new(
            self._challenge_hmac_key,
            f"{challenge_id}:{code}".encode(),
            hashlib.sha256,
        ).hexdigest()

    def _digest_session(self, token: str) -> str:
        return hmac.new(
            self._session_hmac_key, token.encode(), hashlib.sha256
        ).hexdigest()

    def _now(self) -> datetime:
        now = self._clock()
        if now.utcoffset() is None:
            raise RuntimeError("identity clock must return a timezone-aware instant")
        return now


def normalize_email(email: str) -> tuple[str, str]:
    normalized = email.strip().casefold()
    local, separator, domain = normalized.rpartition("@")
    if separator == "" or local == "" or domain == "":
        raise UniversityDomainError("email is not eligible for participation")
    return normalized, domain


def normalize_username(username: str) -> str:
    if USERNAME_PATTERN.fullmatch(username) is None:
        raise UsernameUnavailableError(
            "username must contain 3-24 letters, numbers, or underscores"
        )
    return username.casefold()

File: api/_src/modules/identity/service.ts
/**
 * Passwordless activation, sessions, usernames, and principals
 * (port of platform_app.modules.identity.service).
 */
import { randomUUID } from "node:crypto";

import type { RoleName } from "@energy/shared";
import type { PrismaClient } from "@prisma/client";

import {
  digestChallenge,
  digestSession,
  digestsEqual,
  generateCode,
  generateToken,
} from "../../crypto.js";
import type { PrismaTransaction } from "../../db.js";
import type {
  UniversityVerificationGateway,
  VerifiedResidenceContract,
} from "../university/contracts.js";
import type { Principal } from "./authorization.js";
import {
  InvalidChallengeError,
  InvalidSessionError,
  RosterIneligibleError,
  UniversityDomainError,
  UsernameUnavailableError,
  ChallengeRateLimitError,
} from "./errors.js";
import type { EmailCodeSender } from "./ports.js";

const USERNAME_PATTERN = /^[A-Za-z0-9_]{3,24}$/;
const CHALLENGE_TTL_MS = 10 * 60 * 1000;
const RATE_WINDOW_MS = 15 * 60 * 1000;
const RATE_LIMIT = 3;
const SESSION_TTL_MS = 60 * 60 * 1000;

export interface ChallengeIssued {
  challengeId: string;
  expiresAt: Date;
}

export interface ActivatedSession {
  accessToken: string;
  expiresAt: Date;
  username: string;
  roles: RoleName[];
}

export interface IdentityServiceOptions {
  clock?: () => Date;
  codeFactory?: () => string;
  tokenFactory?: () => string;
}

export class IdentityService {
  private readonly clock: () => Date;
  private readonly codeFactory: () => string;
  private readonly tokenFactory: () => string;

  constructor(
    private readonly db: PrismaClient,
    private readonly gateway: UniversityVerificationGateway,
    private readonly emailSender: EmailCodeSender,
    private readonly challengeHmacKey: string,
    private readonly sessionHmacKey: string,
    options: IdentityServiceOptions = {},
  ) {
    this.clock = options.clock ?? (() => new Date());
    this.codeFactory = options.codeFactory ?? generateCode;
    this.tokenFactory = options.tokenFactory ?? generateToken;
  }

  async requestChallenge(email: string): Promise<ChallengeIssued> {
    const [normalizedEmail, domain] = normalizeEmail(email);
    const university = await this.db.university.findFirst({
      where: {
        status: "active",
        emailDomains: { some: { normalizedDomain: domain } },
      },
    });
    if (university === null || university.rosterReference === null) {
      throw new UniversityDomainError("email is not eligible for participation");
    }

    const now = this.clock();
    const recentCount = await this.db.emailChallenge.count({
      where: {
        normalizedEmail,
        createdAt: { gte: new Date(now.getTime() - RATE_WINDOW_MS) },
      },
    });
    if (recentCount >= RATE_LIMIT) {
      throw new ChallengeRateLimitError("too many verification requests");
    }

    const challengeId = randomUUID();
    const code = this.codeFactory();
    const expiresAt = new Date(now.getTime() + CHALLENGE_TTL_MS);
    await this.db.emailChallenge.create({
      data: {
        id: challengeId,
        universityId: university.id,
        normalizedEmail,
        codeDigest: digestChallenge(this.challengeHmacKey, challengeId, code),
        expiresAt,
        attempts: 0,
        maxAttempts: 5,
        createdAt: now,
      },
    });
    await this.emailSender.sendCode(normalizedEmail, code, expiresAt);
    return { challengeId, expiresAt };
  }

  async verifyChallenge(
    challengeId: string,
    code: string,
    username: string,
  ): Promise<ActivatedSession> {
    const now = this.clock();
    const challenge = await this.db.emailChallenge.findUnique({
      where: { id: challengeId },
    });
    if (
      challenge === null ||
      challenge.consumedAt !== null ||
      challenge.expiresAt.getTime() <= now.getTime() ||
      challenge.attempts >= challenge.maxAttempts
    ) {
      throw new InvalidChallengeError("verification code is invalid");
    }

    const expectedDigest = digestChallenge(
      this.challengeHmacKey,
      challenge.id,
      code,
    );
    if (!digestsEqual(challenge.codeDigest, expectedDigest)) {
      // Persist the failed attempt even though we reject the request.
      await this.db.emailChallenge.update({
        where: { id: challenge.id },
        data: { attempts: { increment: 1 } },
      });
      throw new InvalidChallengeError("verification code is invalid");
    }

    const verification = await this.gateway.verifyResident(
      challenge.normalizedEmail,
      now,
    );
    const university = await this.db.university.findUnique({
      where: { id: challenge.universityId },
    });
    if (
      university === null ||
      verification.status !== "active" ||
      verification.residence === null ||
      verification.studentReference === null ||
      verification.universityReference !== university.rosterReference
    ) {
      await this.db.emailChallenge.update({
        where: { id: challenge.id },
        data: { consumedAt: now },
      });
      throw new RosterIneligibleError(
        "active enrolment and residence are required",
      );
    }

    const normalizedUsername = normalizeUsername(username);
    const studentReference = verification.studentReference;
    const residence = verification.residence;
    const universityId = university.id;

    // Atomic activation: on a username conflict the whole account/identity
    // creation rolls back (matching the original session.rollback() on 409).
    return this.db.$transaction(async (tx) => {
      let identity = await tx.universityIdentity.findFirst({
        where: { universityId, normalizedEmail: challenge.normalizedEmail },
      });
      let accountId: string;
      if (identity === null) {
        const account = await tx.userAccount.create({
          data: { status: "active" },
        });
        accountId = account.id;
        identity = await tx.universityIdentity.create({
          data: {
            universityId,
            accountId: account.id,
            normalizedEmail: challenge.normalizedEmail,
            externalStudentReference: studentReference,
            enrollmentState: "active",
          },
        });
      } else {
        const existingAccount = await tx.userAccount.findUnique({
          where: { id: identity.accountId },
        });
        if (existingAccount === null || existingAccount.status !== "active") {
          throw new RosterIneligibleError("account is not active");
        }
        accountId = existingAccount.id;
        identity = await tx.universityIdentity.update({
          where: { id: identity.id },
          data: {
            externalStudentReference: studentReference,
            enrollmentState: "active",
          },
        });
      }

      const profile = await upsertProfile(
        tx,
        identity.id,
        universityId,
        username,
        normalizedUsername,
      );
      await upsertParticipantRole(tx, accountId, universityId);
      await recordResidence(tx, identity.id, universityId, residence, now);
      await tx.emailChallenge.update({
        where: { id: challenge.id },
        data: { consumedAt: now },
      });

      const token = this.tokenFactory();
      const expiresAt = new Date(now.getTime() + SESSION_TTL_MS);
      await tx.accessSession.create({
        data: {
          accountId,
          tokenDigest: digestSession(this.sessionHmacKey, token),
          expiresAt,
          createdAt: now,
        },
      });

      return {
        accessToken: token,
        expiresAt,
        username: profile.username,
        roles: ["participant"] satisfies RoleName[],
      };
    });
  }

  async principalForToken(token: string): Promise<Principal> {
    const now = this.clock();
    const accessSession = await this.db.accessSession.findFirst({
      where: {
        tokenDigest: digestSession(this.sessionHmacKey, token),
        revokedAt: null,
        expiresAt: { gt: now },
      },
    });
    if (accessSession === null) {
      throw new InvalidSessionError("access token is invalid");
    }
    const assignments = await this.db.roleAssignment.findMany({
      where: { accountId: accessSession.accountId },
    });
    return {
      accountId: accessSession.accountId,
      grants: assignments.map((row) => ({
        role: row.role,
        universityId: row.universityId,
        buildingId: row.buildingId,
      })),
    };
  }

  async changeUsername(principal: Principal, username: string): Promise<string> {
    const normalized = normalizeUsername(username);
    const identity = await this.db.universityIdentity.findFirst({
      where: { accountId: principal.accountId },
    });
    if (identity === null) {
      throw new InvalidSessionError("participant identity is missing");
    }
    const profile = await this.db.userProfile.findFirst({
      where: { identityId: identity.id },
    });
    if (profile === null) {
      throw new InvalidSessionError("participant profile is missing");
    }
    const existing = await this.db.userProfile.findFirst({
      where: {
        universityId: identity.universityId,
        normalizedUsername: normalized,
        id: { not: profile.id },
      },
    });
    if (existing !== null) {
      throw new UsernameUnavailableError("username is unavailable");
    }
    const updated = await this.db.userProfile.update({
      where: { id: profile.id },
      data: { username, normalizedUsername: normalized },
    });
    return updated.username;
  }
}

async function upsertProfile(
  tx: PrismaTransaction,
  identityId: string,
  universityId: string,
  username: string,
  normalizedUsername: string,
) {
  const conflict = await tx.userProfile.findFirst({
    where: {
      universityId,
      normalizedUsername,
      identityId: { not: identityId },
    },
  });
  if (conflict !== null) {
    throw new UsernameUnavailableError("username is unavailable");
  }
  const existing = await tx.userProfile.findUnique({ where: { identityId } });
  if (existing === null) {
    return tx.userProfile.create({
      data: {
        universityId,
        identityId,
        username,
        normalizedUsername,
        moderationState: "active",
      },
    });
  }
  return tx.userProfile.update({
    where: { id: existing.id },
    data: { username, normalizedUsername },
  });
}

async function upsertParticipantRole(
  tx: PrismaTransaction,
  accountId: string,
  universityId: string,
): Promise<void> {
  const existing = await tx.roleAssignment.findFirst({
    where: { accountId, role: "participant", universityId },
  });
  if (existing === null) {
    await tx.roleAssignment.create({
      data: { accountId, role: "participant", universityId, buildingId: null },
    });
  }
}

async function recordResidence(
  tx: PrismaTransaction,
  identityId: string,
  universityId: string,
  residence: VerifiedResidenceContract,
  now: Date,
): Promise<void> {
  const current = await tx.verifiedResidence.findFirst({
    where: { identityId, effectiveEnd: null },
  });
  if (current !== null) {
    const unchanged =
      current.buildingReference === residence.buildingReference &&
      current.apartmentReference === residence.apartmentReference &&
      current.roomReference === residence.roomReference &&
      current.sourceVersion === residence.sourceVersion;
    if (unchanged) {
      await tx.verifiedResidence.update({
        where: { id: current.id },
        data: { verifiedAt: now },
      });
      return;
    }
    await tx.verifiedResidence.update({
      where: { id: current.id },
      data: { effectiveEnd: now },
    });
  }
  await tx.verifiedResidence.create({
    data: {
      universityId,
      identityId,
      buildingReference: residence.buildingReference,
      apartmentReference: residence.apartmentReference,
      roomReference: residence.roomReference,
      sourceVersion: residence.sourceVersion,
      effectiveStart: now,
      effectiveEnd: null,
      verifiedAt: now,
    },
  });
}

export function normalizeEmail(email: string): [string, string] {
  const normalized = email.trim().toLowerCase();
  const separatorIndex = normalized.lastIndexOf("@");
  if (separatorIndex <= 0 || separatorIndex >= normalized.length - 1) {
    throw new UniversityDomainError("email is not eligible for participation");
  }
  const domain = normalized.slice(separatorIndex + 1);
  return [normalized, domain];
}

export function normalizeUsername(username: string): string {
  if (!USERNAME_PATTERN.test(username)) {
    throw new UsernameUnavailableError(
      "username must contain 3-24 letters, numbers, or underscores",
    );
  }
  return username.toLowerCase();
}
